### ISSUE TITLE
Remove deprecated language from `dataIdFromObject` 

### ISSUE THREAD

**Issue by jpvajda** [apollographql/apollo-client#9817]
### Related Issues

https://github.com/apollographql/apollo-client/issues/9799
https://github.com/apollographql/apollo-client/issues/6724

### Issue

In our documentation for [dataIDFromObject](https://www.apollographql.com/docs/react/caching/cache-configuration/#dataidfromobject)  We specify this function will be deprecated soon we haven't decide on when to officially remove this functionality which has lead to some questions from our community.

### History

We think type policies are a safer way to customize cache keys, as outlined in the last few bullet points of our [Customizing identifier generation globally](https://www.apollographql.com/docs/react/caching/cache-configuration/#customizing-identifier-generation-globally) docs. The common `dataIdFromObject` approach is:

-     Sensitive to aliasing mistakes
-     Usually fails to protect against undefined object properties
-     Makes it easy to accidentally use different key fields at different times leading to cache inconsistencies

Therefore it has marked as deprecated for a while.

### Decision

We have decided to not deprecate this functionality as:

`dataIdFromObject` still provides an essential service of identifying objects whose type policies don't have keyFields configured, as a sort of catch-all or backstop for the cache identification system.

`dataIdFromObject` plays a role in how we implement our default support for id-based identification: [defaultDataIdFromObject](https://github.com/apollographql/apollo-client/blob/7c00b03a3678773067aac1ca88d3081030559da0/src/cache/inmemory/helpers.ts#L27-L47).

So we want to update our documentation accordingly.

**Comment by benjamn** [comment]
@jpvajda That may all be true, but `dataIdFromObject` still provides an essential service of identifying objects whose type policies don't have `keyFields` configured, as a sort of catch-all or backstop for the cache identification system.

In fact, `dataIdFromObject` plays a role in how we implement our default support for `id`-based identification: [`defaultDataIdFromObject`](https://github.com/apollographql/apollo-client/blob/7c00b03a3678773067aac1ca88d3081030559da0/src/cache/inmemory/helpers.ts#L27-L47).

In other words, I think this issue gets it backwards: we need to remove language about the "deprecation" of `dataIdFromObject` from our documentation, and not continue with the deprecation at all.

**Comment by jpvajda** [comment]
@benjamn thanks for clarifying! I can update the docs accordingly.
